azkaban-common/src/main/java/azkaban/database/DataSourceUtils.java

```diff
@@ -142,6 +146,45 @@ private MySQLBasicDataSource(String host, int port, String dbName,
       setTestOnBorrow(true);
     }
 
+    /**
+     * This method overrides {@link BasicDataSource#getConnection()}, in order to have retry logics.
+     *
+     */
+    @Override
+    public Connection getConnection() throws SQLException {
+
+      /*
+       * when az server process launches, if the connection isn't valid (e.g., network problem, wrong password, etc),
+       * we fail the connection immediately.
+       */
+      if (getDataSource() == null) {
+        return createDataSource().getConnection();
+      }
+
+      Connection connection = null;
+      int retryAttempt = 0;
+      this.dataSource = null;
+      while (connection == null && retryAttempt < ServerInternals.MAX_DB_RETRY_COUNT) {
+        try {
+          /*
+           * when DB connection could not be fetched here, dbcp library will keep searching until a timeout defined in
+           * its code hardly.
+           */
+          connection = createDataSource().getConnection();
+          if(connection != null)
+            return connection;
+        } catch (Exception ex) {
+          logger.error("Failed to find DB connection", ex);
+          this.dataSource = null;
+          logger.info("waits 10 seconds and retry. No.Attempt = " + retryAttempt);
+//          sleepMillis(1000L);
```

Comment:
Will considering how to add test after the code is reviewed to be valid and efficient.